### PR TITLE
Readme src what and purpose

### PR DESCRIPTION
**Pull request by peter-michalski**
This PR includes infrastructure for adding the 'What' section and 'Purpose' subsection in the generated src README files. 

Example-specific 'What' and 'Purpose' will be added on a case by case basis in other PRs.

The PR stems from issue #3159

### CODE REVIEW

**Review by smiths — CHANGES_REQUESTED** [review]
@JacquesCarette will review the changes to the Drasil code, but as usual I've looked at the changes to stable.  I'm sure  you just intend it as a placeholder, but in some of the case study examples you use an excerpt from the SRS, like for [GlassBR](https://github.com/JacquesCarette/Drasil/pull/3225/files#diff-71afaf9ae4fc29ad7cec3e46d08ec9b86ecdb1739398fd4e1bcdacdcf2d5caef).  Even on a temporary basis, I don't think we should do this.  It is a step in the wrong direction.  Can you change these examples to just have a simple statement of the purpose, or leave them blank, as for the other examples?

**Comment by peter-michalski** [comment]
@smiths Your assumption about the placeholder is correct. I have applied the requested change (leaving purpose blank) in fd6778542ced9ce09a8e964f84225d93ae9356b4.

**Review by smiths — APPROVED** [review]

**Comment by smiths** [comment]
I've approved the changes from my perspective, but @JacquesCarette still has to review.  I've explicitly added @JacquesCarette as a reviewer, but unfortunately the PR is still shown as approved.  Please do not merge until Jacques has reviewed.

**Comment by JacquesCarette** [comment]
[Drasil PR #3225](https://trello.com/c/NZrwBLzM/276-drasil-pr-3225)

File: code/drasil-code/lib/Language/Drasil/Code/Imperative/WriteReadMe.hs [hunk 1]
```diff
@@ -33,3 +36,9 @@ makeReadMe ReadMeInfo {
 makeInstr :: ImplementationType -> [FilePath]-> Doc
 makeInstr Library _ = empty
 makeInstr Program cfp = instDoc cfp
+
+-- |Helper that checks if the field is empty; hels with allowing optional content
```
Comment (JacquesCarette) [inline code/drasil-code/lib/Language/Drasil/Code/Imperative/WriteReadMe.hs:40]: fix typo in comment

File: code/drasil-code/lib/Language/Drasil/Code/Imperative/WriteReadMe.hs [hunk 2]
```diff
@@ -33,3 +36,9 @@ makeReadMe ReadMeInfo {
 makeInstr :: ImplementationType -> [FilePath]-> Doc
 makeInstr Library _ = empty
 makeInstr Program cfp = instDoc cfp
+
+-- |Helper that checks if the field is empty; hels with allowing optional content 
+-- rendering in a README file
+fieldEmptySTR :: String -> Maybe String
```
Comment (JacquesCarette) [inline code/drasil-code/lib/Language/Drasil/Code/Imperative/WriteReadMe.hs:42]: Surely such a function already exists?

File: code/stable/dblpendulum/src/cpp/README.md [hunk 3]
```diff
@@ -1,6 +1,8 @@
 # DblPendulum 
 > Author: Dong Chen
 
+
```
Comment (JacquesCarette) [inline code/stable/dblpendulum/src/cpp/README.md:4]: Something's gone weird: our process shouldn't be adding extra blank lines when there is nothing to show. This is definitely a bug that needs fixed.

**Review by JacquesCarette — CHANGES_REQUESTED** [review]

File: code/drasil-code/lib/Language/Drasil/Code/Imperative/WriteReadMe.hs (hunk 2 above)
Comment (peter-michalski) [reply code/drasil-code/lib/Language/Drasil/Code/Imperative/WriteReadMe.hs]: updated in 24b0ddd47c50a8c21e91729ab7462e6da01d7e37

File: code/drasil-code/lib/Language/Drasil/Code/Imperative/WriteReadMe.hs (hunk 1 above)
Comment (peter-michalski) [reply code/drasil-code/lib/Language/Drasil/Code/Imperative/WriteReadMe.hs]: removed in 24b0ddd47c50a8c21e91729ab7462e6da01d7e37

File: code/stable/dblpendulum/src/cpp/README.md (hunk 3 above)
Comment (peter-michalski) [reply code/stable/dblpendulum/src/cpp/README.md]: fixed in 47b7f219572469885d41258ac508cfeae556cbd8, 454828b0ffc1b2bfcdee74d5be725c14dc469bf1, and 3de86113edf2b71994fc0276c1177d81d05170ba

**Review by JacquesCarette — APPROVED** [review]

**Pull request merged**